assets/src/edit-story/app/history/historyProvider.js

```diff
@@ -40,23 +40,43 @@ function HistoryProvider({ children, size }) {
     versionNumber,
   } = useHistoryReducer(size);
 
+  const [hasNewChanges, setHasNewChanges] = useState(false);
   const setPreventUnload = usePreventWindowUnload();
+  const savedVersionNumber = useRef(false);
 
   useEffect(() => {
-    setPreventUnload('history', versionNumber - 1 > 0);
-
+    setPreventUnload('history', hasNewChanges);
     return () => setPreventUnload('history', false);
-  }, [setPreventUnload, versionNumber]);
+  }, [setPreventUnload, hasNewChanges]);
+
+  useEffect(() => {
+    // The first version number is 1. There are new changes if:
+    // a) If the story hasn't been saved yet and the version number is larger than 1 or,
+    // b) The version number is different from the last saved version number.
+    setHasNewChanges(
```

Comment:
Hmm. Would it be very wrong to simply always increment `versionNumber` on replay? It'd certainly be easier. And I feel like semantically it's not so bad.

Reply:
I might be missing something but how would we then know which is the moment in history that's saved?

E.g.
1. User does something
2. User Saves
3. User clicks Undo
4. User clicks Redo 

How would we detect in step 4. that there are no new changes?

Reply:
Hmm. That's a good point. But my point still stands however: it feels like this functionality better belongs in the reducer since on any dispatch we can know exactly what that state is. What if we did this slightly differently. What if we introduced `lastSaveVersionNumber` in the history reducer's state, right next to the `versionNumber`? Then `hasNewChanges = (versionNumber === lastSaveVersionNumber)`. Right?

Reply:
In this case, the undo and redo still couldn't just increment the `versionNumber` -- otherwise `lastSaveVersionNumber` wouldn't make much sense since the version number would never be restored if it's always only incremented. Or did you mean using `lastSavedVersionNumber` unrelated to incrementing?

It seems to me that the history reducer should not concern itself with the knowledge about when the state in the app was saved -- if we'd track the saved version number inside the history reducer, then we'd have to create a new action in the reducer which would update the history when saving just to mark the saved version number, making the history less independent. Currently, saving per se doesn't trigger any changes in history.

I'm also not convinced about storing the `lastSavedVersionNumber` since it doesn't seem like a generally usable value -- e.g. we can't pull out the last saved version from the history based on that number -- it seems to be a modification to the history based on our specific need instead of adding a generally usable logic which could be applied in any app.

An alternative to handling the version numbers for detecting new changes would be just storing the saved state fully upon every save somewhere and then comparing it with the current state to check if there are differences or not. This way it would also be possible to restore the saved state easily or compare the saved state to a revision or current changes (something we might want to do in the future). Thoughts?

cc @barklund -- perhaps you also have thoughts to share here.

Reply:
@miina These are all good points. I think your current approach is rather similar - you just opting for storing this value in ref and I agree this is better. I think mostly got confused by the complexity of the expression here for `setHasNewChanges(...)`. We could possibly simplify it like this:

```
const savedVersionNumber = useRef(versionNumber); 
useEffect(() => {
  setHasNewVersion(versionNumber !== savedVersionNumber.current);
}, [versionNumber]);
```

I.e. the only thing that really changes here is the initializing value for the `savedVersionNumber` ref. Will this work?

Reply:
Hmm, looks much simpler indeed :) 
It should work but with setting the initial saved version number to `1` instead o `versionNumber`, the reason being that when history is initiated, the `versionNumber` is `0`, it's changed to `1` once the first entry -- the loaded state of the story -- gets added.

Changed in fa9b210c